src/Pinch/Protocol/Internal.hs

```diff
@@ -0,0 +1,37 @@
+module Pinch.Protocol.Internal
+  ( guardMaxBinaryLength
+  , guardMaxMethodNameLength
+  , guardMaxListLength
+  , guardMaxSetSize
+  , guardMaxMapSize
+  , guardPositiveSize
+  ) where
+
+import Pinch.Protocol.Options (ProtocolOptions(..))
+
+import qualified Data.Serialize.Get as G
+
+guardLimit :: (ProtocolOptions -> Int) -> String -> ProtocolOptions -> Int -> G.Get ()
+guardLimit limitSelector measurementName opts n
+  | n > limitSelector opts = fail $ measurementName ++ " exceeds maximum: " ++ show n
+  | otherwise = pure ()
+
+guardMaxBinaryLength :: ProtocolOptions -> Int -> G.Get ()
+guardMaxBinaryLength = guardLimit poMaxBinaryLength "Binary blob length"
+
+guardMaxMethodNameLength :: ProtocolOptions -> Int -> G.Get ()
+guardMaxMethodNameLength = guardLimit poMaxMethodNameLength "Method name length"
+
+guardMaxListLength :: ProtocolOptions -> Int -> G.Get ()
+guardMaxListLength = guardLimit poMaxListLength "List length"
+
+guardMaxSetSize :: ProtocolOptions -> Int -> G.Get ()
+guardMaxSetSize = guardLimit poMaxSetSize "Set size"
+
+guardMaxMapSize :: ProtocolOptions -> Int -> G.Get ()
+guardMaxMapSize = guardLimit poMaxMapSize "Map size"
+
+guardPositiveSize :: String -> Int -> G.Get ()
```

Comment:
suggestion: call this `guardNonNegativeSize`. "positive"  implies `> 0`.

Reply:
Good point...